packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +422,33 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function hasValidReturnType(
+      node:
+        | TSESTree.FunctionExpression
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | null,
+    ): boolean {
+      if (node != null) {
+        if (
+          (node.returnType &&
+            node.returnType.typeAnnotation.type !==
+              AST_NODE_TYPES.TSAnyKeyword &&
+            node.returnType.typeAnnotation.type !==
+              AST_NODE_TYPES.TSUnknownKeyword) ||
```

Comment:
[Bug] I somehow missed this in previous reviews: we can't rely on syntactic check here

```ts
(): any => console.log(); // reported - nice!
(): unknown => console.log(); // reported - nice!

type Foo = any
(): Foo => console.log(); // not reported - bug!
type Bar = unknown
(): Bar => console.log(); // not reported - bug!
```

[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.ts&code=BQSgXABAhgdgnhAvAPggYwPYwM4YDYCmAdHhgOagDcEA9DRAE4EAOGDALgQCYQC0EMAJZoCAQgBQoSAFcYAaxgYA7jCSpMOfMVIUQ1OoxZtOPfkJETx7OMwIQAYhgxJo8SeAdO16LLkIlyKlp6RXZDVg5uPggAI2kyCWtbCAAhKAYXWQVlGHdINIyUH01-HSCDUPDjKP44hKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANrhsORNGgdokADSatWLdkiVhnRQDVJUgJJN701fjiIT2AL7eAupp%2BPkA&tsconfig=&tokens=false)

Reply:
When specifying a type using a type alias, it is difficult to find a way to determine what type it is.

Could you please give me a hint?

```
type Bar = unknown
(): Bar => console.log();  // how to know Bar is unknown type in create fn
```

Reply:
@auvred Can you help with the issue left in the comment above?

Reply:
I'm not 100% sure about this, but I'd try playing around with function signatures. There is nice utility function in the `ts-api-utils` package - `getCallSignaturesOfType`.

See: https://github.com/JoshuaKGoldberg/ts-api-utils/blob/d808eba09baa3ba093cd31b85aad57bdcc539867/src/types/getters.ts#L23

Search with some examples of usage: https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint%20getCallSignaturesOfType&type=code

Reply:
it works to me very well. thank you!